Prevent overshoot in rotateToHeading()

Tacha-S opened the pull request:
- fix #18

Tacha-S committed:
Prevent overshoot for final rotation

arthurgomes4 commented:
hey @Tacha-S thank you for submitting this request. Its a good catch but your solution needs a bit of work. 

1. Im not sure you're using the right calculation for `angle_to_stop`. Check that once please.
2. I dont think assigning the minimum feasible velocity is a good braking method because it seems like that may lead to oscillations in the commanded velocity (between max feasible and min feasible velocity). 

But Id like some clarification first, are you facing this issue in the final rotation to target or when rotating to align with a path?

Tacha-S commented:
It happens during the final rotation to the target.

I believe the calculation for `angle_to_stop` is correct since it's based on the equation **½ a t² + v₀ t**.

arthurgomes4 commented:
Yes it is, but the assignment statement for angle_to_stop effectively works out to:

`( 0.5 * a * t^2 + u*t ) + u*t`

Given that the sign variable will never be zero. 

Why is there an extra `u*t` here? (u is the initial velocity).

Also, you could create a separate function for the final rotation to target orientation. I don't think this braking is required for the initial rotation to path.

Tacha-S commented:
Ah, sorry about that — I forgot to remove the first `u*t` when simplifying the equation. I’ll fix that.

You’re right, my adjustment currently applies only to the final rotation, so I’ll separate it into a dedicated function for that phase.
However, if `rotate_to_heading_min_angle` is small and the rotation speed is high, this issue might also occur during the initial rotation.

Tacha-S commented:
@arthurgomes4 I fixed it. Anything else?

arthurgomes4 commented:
alright it looks good to me, great work @Tacha-S . Can you squash your commits into one?

Tacha-S commented:
Sure. I’ve squashed my commits into one.

arthurgomes4 commented:
@Tacha-S one last question before we merge this: Why have you used a ternary operator here?

```cpp
  const double sign = angle_to_path > 0.0 ? 1.0 : -1.0;
  const double time_to_stop = std::abs(angular_vel) / max_angular_accel_;
  const double angle_to_stop = (sign ? 0.5 * max_angular_accel_ * std::pow(time_to_stop, 2)
                                     : -0.5 * max_angular_accel_ * std::pow(time_to_stop, 2)) +
                               curr_speed.angular.z * time_to_stop;
```

Shouldnt this be: 

```cpp
  const double angle_to_stop = sign * 0.5 * max_angular_accel_ * std::pow(time_to_stop, 2) +
                               curr_speed.angular.z * time_to_stop;
```

Tacha-S commented:
@arthurgomes4 I fixed it.
Also, I had been calculating `time_to_stop` using the commanded `angular_vel`, but I’ve corrected it to use the current velocity instead.
https://github.com/blackcoffeerobotics/vector_pursuit_controller/compare/b888f4cfd32099a75cbc9891fb297b33db982f4a..ff6a1b7f608350ed83bf0542a11dfd09eff42ba5
